File: Python/ByTier/F/1970_F3_Playing_Quidditch_Hard.py
# Problem: CF 1970 F3 - Playing Quidditch (Hard)
# https://codeforces.com/contest/1970/problem/F3

"""
Code Purpose:
This code simulates a Quidditch game based on a given field setup and sequence of actions. 
It tracks player positions, ball movement, scoring, eliminations, and Golden Snitch catches to determine the final score.

Algorithms/Techniques:
- Simulation-based approach to model game state
- Hash map for O(1) lookup of entity positions
- Coordinate movement in grid using direction mapping
- Event-based triggers (scoring, elimination, catching Snitch)
- Sorting for consistent output order of eliminated players

Time Complexity: O(T * N * M) where T is number of steps, N and M are dimensions of grid
Space Complexity: O(N * M) for storing field and position mappings

"""

N, M = map(int, input().split())
field = [input().split() for _ in range(N)]
assert all(len(row) == M for row in field)
pos = {
    ent: [r, c]
    for r, row in enumerate(field)
    for c, ent in enumerate(row)
    if ent not in ["..", "RG", "BG"]  # Exclude empty cells and goals
}
score = [0, 0]  # [Red score, Blue score]
T = int(input())
for t in range(T):
    ent, arg, *opt = input().split()
    if arg in "UDLR":
        # Process movement action
        d = "UDLR".index(arg)
        pos[ent][d // 2] += [-1, +1][d % 2]  # Move in direction
        # Check if any player is on Bludger's position and get eliminated
        elim = sorted(ent for ent in pos if ent[0] != "." and pos[ent] == pos.get(".B"))
        for ent in elim:
            print(t, ent, "ELIMINATED")
            del pos[ent]  # Remove eliminated player
    elif arg == "T":
        # Process throwing the Quaffle
        r, c = pos[ent]
        g = field[r][c]
        if g[1] == "G":
            # Check if thrown into a goal
            d = "BR".index(g[0])  # Determine which team scored
            print(t, ["RED", "BLUE"][d], "GOAL")
            score[d] += 1
    else:
        # arg == "C" - Catching action
        assert arg == "C"
        [opt] = opt
        if opt == ".S":
            # Catching the Golden Snitch
            d = "RB".index(ent[0])  # Determine which team caught Snitch
            print(t, ["RED", "BLUE"][d], "CATCH GOLDEN SNITCH")
            score[d] += 10  # Award bonus points
print("FINAL SCORE:", *score)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/F/1970_F3_Playing_Quidditch_Hard.cpp
// Problem: CF 1970 F3 - Playing Quidditch (Hard)
// https://codeforces.com/contest/1970/problem/F3

/*
 * Problem: F3. Playing Quidditch (Hard)
 *
 * Purpose: Simulate a Quidditch match with players, balls, goals, and special items.
 *          Determine the winner based on scoring and catching the Golden Snitch.
 *
 * Algorithms/Techniques:
 * - Grid simulation with entity tracking
 * - State management for player carrying Quaffle
 * - Collision detection for elimination and scoring
 * - Game termination on Golden Snitch catch
 *
 * Time Complexity: O(T * N * M) where T is number of steps, and N,M are grid dimensions
 * Space Complexity: O(N * M + T) for grid storage and actions
 */

#include <stdlib.h>
#include <algorithm>
#include <array>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
using V = array<int, 2>;
int n, m, T, bel, rs, bs;
V qua, blu, gs, b[10], r[10];
vector<V> bg, rg;

// Move entity in specified direction
void mov(V &v, char c) {
  if (c == 'U')
    v[0]--;
  if (c == 'D')
    v[0]++;
  if (c == 'L')
    v[1]--;
  if (c == 'R')
    v[1]++;
}

// Print final score and exit
void print() {
  cout << "FINAL SCORE: " << rs << " " << bs;
  exit(0);
}

int main() {
  cin.tie(0)->sync_with_stdio(0);
  cin >> n >> m;
  for (int i = 1; i <= n; i++)
    for (int j = 1; j <= m; j++) {
      string op;
      cin >> op;
      if (op == "BG")
        bg.push_back({i, j});
      else if (op[0] == 'B')
        b[op[1] - '0'] = {i, j};
      if (op == "RG")
        rg.push_back({i, j});
      else if (op[0] == 'R')
        r[op[1] - '0'] = {i, j};
      if (op == ".Q")
        qua = {i, j};
      if (op == ".B")
        blu = {i, j};
      if (op == ".S")
        gs = {i, j};
    }
  cin >> T;
  for (int t = 0; t < T; t++) {
    string p, q;
    char c;
    cin >> p >> c;
    if (c == 'C') {
      cin >> q;
      if (q == ".S") {
        cout << t << (p[0] == 'R' ? " RED" : " BLUE")
             << " CATCH GOLDEN SNITCH\n";
        (p[0] == 'R' ? rs : bs) += 10;
        print();
      }
      if (q == ".Q")
        bel = (p[0] == 'R') * 10 + p[1] - '0' + 1;
    } else if (c == 'T') {
      bel = 0;
      for (auto &x : bg)
        if (x == qua)
          cout << t << " RED GOAL\n", qua = {(n + 1) / 2, (m + 1) / 2}, rs++;
      for (auto &x : rg)
        if (x == qua)
          cout << t << " BLUE GOAL\n", qua = {(n + 1) / 2, (m + 1) / 2}, bs++;
    } else if (p == ".S")
      mov(gs, c);
    else if (p == ".Q")
      mov(qua, c);
    else if (p == ".B") {
      mov(blu, c);
      for (int i = 0; i < 10; i++)
        if (b[i] == blu) {
          cout << t << " B" << i << " ELIMINATED\n", b[i] = {0, 0};
          if (bel == i + 1)
            bel = 0;
        }
      for (int i = 0; i < 10; i++)
        if (r[i] == blu) {
          cout << t << " R" << i << " ELIMINATED\n", r[i] = {0, 0};
          if (bel == i + 11)
            bel = 0;
        }
    } else {
      int id = (p[0] == 'R') * 10 + p[1] - '0' + 1;
      V &a = id > 10 ? r[id - 11] : b[id - 1];
      mov(a, c);
      if (bel == id)
        qua = a;
      if (a == blu) {
        cout << t << " " << p << " ELIMINATED\n", a = {0, 0};
        if (bel == id)
          bel = 0;
      }
    }
  }
  print();
}


// https://github.com/VaHiX/CodeForces/